File: android/app/src/main/java/dev/ghostty/connect/terminal/ItermInlineImageParser.kt
package dev.ghostty.connect.terminal

import java.io.ByteArrayOutputStream
import java.util.Base64

internal class ItermInlineImageParser(
    private val onBytes: (ByteArray) -> Unit,
    private val onImage: (ItermInlineImage) -> Unit,
) {
    private val plain = ByteArrayOutputStream()
    private val control = ByteArrayOutputStream()
    private var prefixIndex = 0
    private var capturing = false
    private var pendingEscape = false
    private var draining = false
    private var multipart: Multipart? = null

    fun feed(bytes: ByteArray) {
        bytes.forEach(::accept)
        flushPlain()
    }

    fun reset() {
        plain.reset()
        control.reset()
        prefixIndex = 0
        capturing = false
        pendingEscape = false
        draining = false
        multipart = null
    }

    private fun accept(value: Byte) {
        if (capturing) {
            when {
                pendingEscape && value == ST_END -> finishControl(byteArrayOf(ESC, ST_END))
                pendingEscape -> {
                    appendControl(ESC)
                    pendingEscape = false
                    when (value) {
                        BEL -> finishControl(byteArrayOf(BEL))
                        ESC -> pendingEscape = true
                        else -> appendControl(value)
                    }
                }
                value == BEL -> finishControl(byteArrayOf(BEL))
                value == ESC -> pendingEscape = true
                else -> appendControl(value)
            }
            return
        }

        if (value == PREFIX[prefixIndex]) {
            prefixIndex++
            if (prefixIndex == PREFIX.size) {
                flushPlain()
                prefixIndex = 0
                capturing = true
                control.reset()
            }
            return
        }

        if (prefixIndex > 0) {
            plain.write(PREFIX, 0, prefixIndex)
            prefixIndex = 0
            if (value == PREFIX[0]) {
                prefixIndex = 1
                return
            }
        }
        plain.write(value.toInt())
    }

    private fun appendControl(value: Byte) {
        if (draining) return
        if (control.size() >= MAX_CONTROL_BYTES) {
            draining = true
            control.reset()
            multipart = null
            return
        }
        control.write(value.toInt())
    }

    private fun finishControl(terminator: ByteArray) {
        if (!draining) {
            val content = control.toByteArray()
            if (!handleControl(content)) {
                plain.write(PREFIX)
                plain.write(content)
                plain.write(terminator)
            }
        }
        control.reset()
        capturing = false
        pendingEscape = false
        draining = false
    }

    private fun handleControl(content: ByteArray): Boolean {
        return when {
            content.startsWith(FILE) -> {
                val separator = (FILE.size until content.size).firstOrNull { content[it] == COLON } ?: -1
                if (separator < 0 || separator > MAX_HEADER_BYTES) return true
                val options = parseOptions(content.copyOfRange(FILE.size, separator)) ?: return true
                complete(options, content.copyOfRange(separator + 1, content.size))
                true
            }
            content.startsWith(MULTIPART_FILE) -> {
                multipart = parseOptions(content.copyOfRange(MULTIPART_FILE.size, content.size))
                    ?.let { Multipart(it) }
                true
            }
            content.startsWith(FILE_PART) -> {
                multipart?.append(content, FILE_PART.size)
                true
            }
            content.contentEquals(FILE_END) -> {
                multipart?.let { complete(it.options, it.encoded.toByteArray()) }
                multipart = null
                true
            }
            else -> false
        }
    }

    private fun complete(options: Map<String, String>, encoded: ByteArray) {
        if (options["inline"] != "1") return
        val declaredSize = options["size"]?.toLongOrNull()
        if (declaredSize != null && declaredSize !in 0..MAX_IMAGE_BYTES.toLong()) return
        val compact = encoded.filterNot { it == '\r'.code.toByte() || it == '\n'.code.toByte() }.toByteArray()
        if (compact.size > MAX_ENCODED_BYTES || compact.any { !it.isBase64Byte() }) return
        val decoded = runCatching { Base64.getDecoder().decode(compact) }.getOrNull() ?: return
        if (decoded.size > MAX_IMAGE_BYTES || declaredSize?.let { it != decoded.size.toLong() } == true) return
        onImage(ItermInlineImage(options, decoded))
    }

    private fun parseOptions(bytes: ByteArray): Map<String, String>? = runCatching {
        if (bytes.size > MAX_HEADER_BYTES) return null
        val text = bytes.toString(Charsets.US_ASCII)
        if (text.isEmpty()) return emptyMap()
        buildMap {
            text.split(';').forEach { argument ->
                val split = argument.indexOf('=')
                require(split > 0)
                val key = argument.substring(0, split)
                require(key in SUPPORTED_OPTIONS && key !in this)
                put(key, argument.substring(split + 1))
            }
        }
    }.getOrNull()

    private fun flushPlain() {
        if (plain.size() == 0) return
        onBytes(plain.toByteArray())
        plain.reset()
    }

    private fun ByteArray.startsWith(prefix: ByteArray): Boolean =
        size >= prefix.size && prefix.indices.all { this[it] == prefix[it] }

    private fun Byte.isBase64Byte(): Boolean {
        val character = toInt().toChar()
        return character in 'A'..'Z' || character in 'a'..'z' || character in '0'..'9' ||
            character == '+' || character == '/' || character == '='
    }

    private inner class Multipart(val options: Map<String, String>) {
        val encoded = ByteArrayOutputStream()

        fun append(content: ByteArray, offset: Int) {
            val length = content.size - offset
            if (length < 0 || encoded.size() > MAX_ENCODED_BYTES - length) {
                multipart = null
                return
            }
            encoded.write(content, offset, length)
        }
    }

    companion object {
        private val PREFIX = "\u001b]1337;".toByteArray()
        private val FILE = "File=".toByteArray()
        private val MULTIPART_FILE = "MultipartFile=".toByteArray()
        private val FILE_PART = "FilePart=".toByteArray()
        private val FILE_END = "FileEnd".toByteArray()
        private val SUPPORTED_OPTIONS = setOf("name", "size", "width", "height", "preserveAspectRatio", "inline")
        private const val MAX_HEADER_BYTES = 4096
        private const val MAX_IMAGE_BYTES = 4 * 1024 * 1024
        private const val MAX_ENCODED_BYTES = 6 * 1024 * 1024
        private const val MAX_CONTROL_BYTES = MAX_ENCODED_BYTES + MAX_HEADER_BYTES
        private const val BEL: Byte = 0x07
        private const val ESC: Byte = 0x1b
        private const val ST_END: Byte = 0x5c
        private const val COLON: Byte = 0x3a
    }
}

internal data class ItermInlineImage(
    val options: Map<String, String>,
    val data: ByteArray,
)

File: ios/GhosttyConnect/Terminal/ItermInlineImageParser.swift
import Foundation

struct ItermInlineImage: Equatable {
    let options: [String: String]
    let data: Data
}

final class ItermInlineImageParser {
    private static let prefix = Array("\u{1b}]1337;".utf8)
    private static let supportedOptions = Set(["name", "size", "width", "height", "preserveAspectRatio", "inline"])
    private static let maxHeaderBytes = 4_096
    private static let maxImageBytes = 4 * 1_024 * 1_024
    private static let maxEncodedBytes = 6 * 1_024 * 1_024

    private let onBytes: (Data) -> Void
    private let onImage: (ItermInlineImage) -> Void
    private var plain: [UInt8] = []
    private var control: [UInt8] = []
    private var prefixIndex = 0
    private var capturing = false
    private var pendingEscape = false
    private var draining = false
    private var multipart: (options: [String: String], encoded: [UInt8])?

    init(onBytes: @escaping (Data) -> Void, onImage: @escaping (ItermInlineImage) -> Void) {
        self.onBytes = onBytes
        self.onImage = onImage
    }

    func feed(_ data: Data) {
        data.forEach(accept)
        flushPlain()
    }

    func reset() {
        plain.removeAll(keepingCapacity: true)
        control.removeAll(keepingCapacity: true)
        prefixIndex = 0
        capturing = false
        pendingEscape = false
        draining = false
        multipart = nil
    }

    private func accept(_ byte: UInt8) {
        if capturing {
            if pendingEscape && byte == 0x5c { finish(terminator: [0x1b, 0x5c]); return }
            if pendingEscape {
                appendControl(0x1b)
                pendingEscape = false
                if byte == 0x07 { finish(terminator: [0x07]) }
                else if byte == 0x1b { pendingEscape = true }
                else { appendControl(byte) }
            } else if byte == 0x07 { finish(terminator: [0x07]) }
            else if byte == 0x1b { pendingEscape = true }
            else { appendControl(byte) }
            return
        }

        if byte == Self.prefix[prefixIndex] {
            prefixIndex += 1
            if prefixIndex == Self.prefix.count {
                flushPlain()
                prefixIndex = 0
                capturing = true
                control.removeAll(keepingCapacity: true)
            }
            return
        }
        if prefixIndex > 0 {
            plain.append(contentsOf: Self.prefix.prefix(prefixIndex))
            prefixIndex = 0
            if byte == Self.prefix[0] { prefixIndex = 1; return }
        }
        plain.append(byte)
    }

    private func appendControl(_ byte: UInt8) {
        guard !draining else { return }
        guard control.count < Self.maxEncodedBytes + Self.maxHeaderBytes else {
            draining = true
            control.removeAll()
            multipart = nil
            return
        }
        control.append(byte)
    }

    private func finish(terminator: [UInt8]) {
        if !draining && !handleControl(control) {
            plain.append(contentsOf: Self.prefix)
            plain.append(contentsOf: control)
            plain.append(contentsOf: terminator)
        }
        control.removeAll(keepingCapacity: true)
        capturing = false
        pendingEscape = false
        draining = false
    }

    private func handleControl(_ bytes: [UInt8]) -> Bool {
        if let remainder = remainder(bytes, after: "File=") {
            guard let separator = remainder.firstIndex(of: 0x3a), separator <= Self.maxHeaderBytes else { return true }
            guard let options = parseOptions(Array(remainder[..<separator])) else { return true }
            complete(options: options, encoded: Array(remainder[remainder.index(after: separator)...]))
            return true
        }
        if let options = remainder(bytes, after: "MultipartFile=").flatMap(parseOptions) {
            multipart = (options, [])
            return true
        }
        if let part = remainder(bytes, after: "FilePart=") {
            guard var current = multipart, current.encoded.count <= Self.maxEncodedBytes - part.count else { multipart = nil; return true }
            current.encoded.append(contentsOf: part)
            multipart = current
            return true
        }
        if bytes == Array("FileEnd".utf8) {
            if let multipart { complete(options: multipart.options, encoded: multipart.encoded) }
            multipart = nil
            return true
        }
        return false
    }

    private func complete(options: [String: String], encoded: [UInt8]) {
        guard options["inline"] == "1" else { return }
        let declared = options["size"].flatMap(Int.init)
        if let declared, !(0...Self.maxImageBytes).contains(declared) { return }
        let compact = encoded.filter { $0 != 0x0a && $0 != 0x0d }
        guard compact.count <= Self.maxEncodedBytes,
              compact.allSatisfy({ byte in
                  (65...90).contains(byte) || (97...122).contains(byte) || (48...57).contains(byte) || byte == 43 || byte == 47 || byte == 61
              }),
              let data = Data(base64Encoded: Data(compact)),
              data.count <= Self.maxImageBytes,
              declared == nil || declared == data.count else { return }
        onImage(ItermInlineImage(options: options, data: data))
    }

    private func parseOptions(_ bytes: [UInt8]) -> [String: String]? {
        guard bytes.count <= Self.maxHeaderBytes, let text = String(bytes: bytes, encoding: .ascii) else { return nil }
        if text.isEmpty { return [:] }
        var options: [String: String] = [:]
        for argument in text.split(separator: ";", omittingEmptySubsequences: false) {
            guard let equals = argument.firstIndex(of: "=") else { return nil }
            let key = String(argument[..<equals])
            guard !key.isEmpty, Self.supportedOptions.contains(key), options[key] == nil else { return nil }
            options[key] = String(argument[argument.index(after: equals)...])
        }
        return options
    }

    private func remainder(_ bytes: [UInt8], after prefix: String) -> [UInt8]? {
        let prefixBytes = Array(prefix.utf8)
        guard bytes.starts(with: prefixBytes) else { return nil }
        return Array(bytes.dropFirst(prefixBytes.count))
    }

    private func flushPlain() {
        guard !plain.isEmpty else { return }
        onBytes(Data(plain))
        plain.removeAll(keepingCapacity: true)
    }
}
